File: buildroot-server/protocol/codec.py
import json
import logging
from typing import Tuple, Optional, Dict, Any
from pydantic import BaseModel, ValidationError

from protocol.constants import MessageType
from protocol.models import (
    BaseMessage,
    RegisterRequest,
    RegisterResult,
    Heartbeat,
    SystemStatus,
    LogUpload,
    ScriptRecv,
    ScriptResult,
    PtyCreate,
    PtyData,
    PtyResize,
    PtyClose,
    FileRequest,
    FileData,
    FileListRequest,
    FileListResponse,
    FileUploadStart,
    FileUploadData,
    FileUploadAck,
    FileUploadComplete,
    FileDownloadRequest,
    FileDownloadData,
    DownloadPackage,
    CmdRequest,
    CmdResponse,
    DeviceList,
    UpdateCheck,
    UpdateInfo,
    UpdateDownload,
    UpdateApprove,
    UpdateProgress,
    UpdateComplete,
    UpdateError,
    UpdateRollback,
)

logger = logging.getLogger(__name__)


class MessageCodec:
    """消息编解码器"""

    MESSAGE_MODEL_MAP: Dict[int, type[BaseModel]] = {
        MessageType.REGISTER: RegisterRequest,
        MessageType.REGISTER_RESULT: RegisterResult,
        MessageType.HEARTBEAT: Heartbeat,
        MessageType.SYSTEM_STATUS: SystemStatus,
        MessageType.LOG_UPLOAD: LogUpload,
        MessageType.SCRIPT_RECV: ScriptRecv,
        MessageType.SCRIPT_RESULT: ScriptResult,
        MessageType.PTY_CREATE: PtyCreate,
        MessageType.PTY_DATA: PtyData,
        MessageType.PTY_RESIZE: PtyResize,
        MessageType.PTY_CLOSE: PtyClose,
        MessageType.FILE_REQUEST: FileRequest,
        MessageType.FILE_DATA: FileData,
        MessageType.FILE_LIST_REQUEST: FileListRequest,
        MessageType.FILE_LIST_RESPONSE: FileListResponse,
        MessageType.FILE_UPLOAD_START: FileUploadStart,
        MessageType.FILE_UPLOAD_DATA: FileUploadData,
        MessageType.FILE_UPLOAD_ACK: FileUploadAck,
        MessageType.FILE_UPLOAD_COMPLETE: FileUploadComplete,
        MessageType.FILE_DOWNLOAD_REQUEST: FileDownloadRequest,
        MessageType.FILE_DOWNLOAD_DATA: FileDownloadData,
        MessageType.DOWNLOAD_PACKAGE: DownloadPackage,
        MessageType.CMD_REQUEST: CmdRequest,
        MessageType.CMD_RESPONSE: CmdResponse,
        MessageType.DEVICE_LIST: DeviceList,
        MessageType.UPDATE_CHECK: UpdateCheck,
        MessageType.UPDATE_INFO: UpdateInfo,
        MessageType.UPDATE_DOWNLOAD: UpdateDownload,
        MessageType.UPDATE_APPROVE: UpdateApprove,
        MessageType.UPDATE_PROGRESS: UpdateProgress,
        MessageType.UPDATE_COMPLETE: UpdateComplete,
        MessageType.UPDATE_ERROR: UpdateError,
        MessageType.UPDATE_ROLLBACK: UpdateRollback,
    }

    @classmethod
    def encode(cls, msg_type: int, data: dict | BaseModel) -> bytes:
        """编码消息"""
        if isinstance(data, BaseModel):
            json_data = data.model_dump(exclude_none=True)
        else:
            json_data = data

        json_bytes = json.dumps(json_data, ensure_ascii=False).encode("utf-8")
        json_len = len(json_bytes)

        msg = bytes([msg_type]) + json_len.to_bytes(2, "big") + json_bytes

        logger.debug(
            f"[CREATE_MSG] type=0x{msg_type:02X}, len={json_len}, hex={msg.hex()[:50]}...{msg.hex()[-30:] if len(msg.hex()) > 80 else ''}"
        )
        return msg

    @classmethod
    def decode(cls, raw_data: bytes) -> Tuple[Optional[int], Optional[dict]]:
        """解码消息"""
        if len(raw_data) < 3:
            return None, None

        msg_type = raw_data[0]
        length_bytes = raw_data[1:3]
        json_len = (length_bytes[0] << 8) | length_bytes[1]

        if len(raw_data) < 3 + json_len:
            logger.warning(
                f"消息不完整: 期望{3 + json_len}字节, 实际{len(raw_data)}字节"
            )
            return msg_type, {}

        json_data_bytes = raw_data[3 : 3 + json_len]

        try:
            json_str = json_data_bytes.decode("utf-8")
            if not json_str.strip():
                return msg_type, {}

            model_class = cls.MESSAGE_MODEL_MAP.get(msg_type, BaseMessage)
            model = model_class.model_validate_json(json_str)
            return msg_type, model.model_dump()

        except (json.JSONDecodeError, UnicodeDecodeError) as e:
            logger.warning(f"消息JSON解析失败: {e}")
            return msg_type, {}
        except ValidationError as e:
            logger.warning(f"消息数据验证失败: {e}")
            return msg_type, {}

File: buildroot-web-vue/src/api/codec.ts
// Protocol Codec - Buildroot Agent Binary Protocol
// 编码/解码二进制协议消息

export class MessageCodec {
  /**
   * 编码消息为二进制格式
   * 格式: [msg_type: 1字节] [length: 2字节大端序] [json_data]
   */
  static encode(msgType: number, data: any): Uint8Array {
    // 将数据转换为 JSON
    const jsonData = typeof data === 'object' ? data : { data }
    const jsonString = JSON.stringify(jsonData)
    const jsonBytes = new TextEncoder().encode(jsonString)
    const jsonLen = jsonBytes.length

    // 创建消息缓冲区: [msg_type] [len_hi] [len_lo] [json_data]
    const buffer = new Uint8Array(3 + jsonLen)
    buffer[0] = msgType
    buffer[1] = (jsonLen >> 8) & 0xFF // 高字节
    buffer[2] = jsonLen & 0xFF         // 低字节
    buffer.set(jsonBytes, 3)

    return buffer
  }

  /**
   * 解码二进制消息
   * 格式: [msg_type: 1字节] [length: 2字节大端序] [json_data]
   */
  static decode(rawData: ArrayBuffer | Uint8Array): { type: number; data: any } | null {
    const buffer = rawData instanceof ArrayBuffer
      ? new Uint8Array(rawData)
      : rawData

    if (buffer.length < 3) {
      console.error('Message too short:', buffer.length)
      return null
    }

    // 解析消息头
    const msgType = buffer[0]
    const jsonLen = (buffer[1] << 8) | buffer[2]

    // 检查长度
    if (buffer.length < 3 + jsonLen) {
      console.error(`Message incomplete: expected ${3 + jsonLen}, got ${buffer.length}`)
      return null
    }

    // 解析 JSON 数据
    const jsonBytes = buffer.slice(3, 3 + jsonLen)
    try {
      const jsonString = new TextDecoder().decode(jsonBytes)
      const data = JSON.parse(jsonString)
      return { type: msgType, data }
    } catch (error) {
      console.error('Failed to decode message:', error)
      return null
    }
  }
}

// 消息类型常量（与后端对应）
export const MessageType = {
  HEARTBEAT: 0x01,
  SYSTEM_STATUS: 0x02,
  LOG_UPLOAD: 0x03,
  SCRIPT_RECV: 0x04,
  SCRIPT_RESULT: 0x05,
  PTY_CREATE: 0x10,
  PTY_DATA: 0x11,
  PTY_RESIZE: 0x12,
  PTY_CLOSE: 0x13,
  FILE_REQUEST: 0x20,
  FILE_DATA: 0x21,
  FILE_LIST_REQUEST: 0x22,
  FILE_LIST_RESPONSE: 0x23,
  DOWNLOAD_PACKAGE: 0x24,
  FILE_DOWNLOAD_REQUEST: 0x25,
  FILE_DOWNLOAD_DATA: 0x26,
  CMD_REQUEST: 0x30,
  CMD_RESPONSE: 0x31,
  DEVICE_LIST: 0x50,
  DEVICE_DISCONNECT: 0x51,
  DEVICE_UPDATE: 0x52,
  REGISTER: 0xF0,
  REGISTER_RESULT: 0xF1,
  UPDATE_CHECK: 0x60,
  UPDATE_INFO: 0x61,
  UPDATE_DOWNLOAD: 0x62,
  UPDATE_PROGRESS: 0x63,
  UPDATE_COMPLETE: 0x65,
  UPDATE_ERROR: 0x66,
  UPDATE_ROLLBACK: 0x67,
  UPDATE_REQUEST_APPROVAL: 0x68,
  UPDATE_DOWNLOAD_READY: 0x69,
  UPDATE_APPROVE_INSTALL: 0x6A,
  UPDATE_DENY: 0x6B,
  UPDATE_APPROVE_DOWNLOAD: 0x6C,
  PING_STATUS: 0x70,
} as const

// 消息类型名称映射
export const MessageTypeNames: Record<number, string> = {}
for (const [name, value] of Object.entries(MessageType)) {
  MessageTypeNames[value as number] = name
}
